sympy/solvers/solveset.py

```diff
@@ -478,10 +478,10 @@ def _solve_as_poly(f, symbol, domain=S.Complexes):
     if result is not None:
         if isinstance(result, FiniteSet):
             # this is to simplify solutions like -sqrt(-I) to sqrt(2)/2
-            # - sqrt(2)*I/2. We are not expanding for solution with free
-            # variables because that makes the solution more complicated. For
-            # example expand_complex(a) returns re(a) + I*im(a)
-            if all([s.free_symbols == set() and not isinstance(s, RootOf)
```

Comment:
`s.atoms(Symbol)` may not be empty even if `s.free_symbols == set()`: https://github.com/sympy/sympy/blob/b678730491350c4af13c570ef3ce21d25c3f6178/sympy/core/basic.py#L485-L496

Reply:
Yes, I'm aware of that. The presence of any kind of symbols should prevent the expression from being sent to expand_complex: free symbols, bound symbols, undefined functions. Only numeric expressions should go there.

Reply:
Good caution by @ylemkimon and good rationale for change given in response by @normalhuman.